### ISSUE TITLE
[Q] ...Set some Curl options from Client/ClientOptions

### ISSUE THREAD

**Issue by spnshah** [googleapis/google-cloud-cpp#6136]
Hello,

I wanted to set some specific CURL options from Client/ClientOptions or some other way. I noticed there is a provision using ClientOptions for things like URL, stall_timeout, max socket recv/send size etc.

But, is there a way to pass CURL handle down to the SDK, or set other CURL options that aren't available as part of ClientOptions? Also, any way to make custom socket callback functions? 

Thanks

**Comment by coryan** [comment]
> But, is there a way to pass CURL handle down to the SDK, or set other CURL options that aren't available as part of ClientOptions? Also, any way to make custom socket callback functions?

The answer is both cases is "no".  If you could tell us about what problem is the problem you are trying to solve, maybe we can come up with some suggestions, or at least it could motivate a feature request.

**Comment by spnshah** [comment]
There are couple of things we are trying to do from our side such as:

1. Enable linger option (SO_LINGER) for the socket and initial effort was to use CURLOPT_SOCKOPTFUNCTION using curl_easy_setopt(). We benefited doing this for our workload for S3 cloud in their SDK and feel that if conditions are similar, we need similar implementation for google cloud and its cpp SDK too.
2. Set custom TCP keep intervals and idle times.
3. Set connection timeouts (CURLOPT_CONNECTTIMEOUT). We have provided our user an option to specify timeouts for other clouds and so we want to consistently honor that for Google cloud too.

So to implement all of these, the options I knew was to get the CURL handle that cloud-cpp SDK is using and add options to it, or create a curl handle and after setting the options, pass it over to SDK to carry it further.

Also, just so you know, we got locked on v0.20.x of google-cloud-cpp last year.

**Comment by coryan** [comment]
> There are couple of things we are trying to do from our side such as:
> 


> 1. Enable linger option (SO_LINGER) for the socket and initial effort was to use CURLOPT_SOCKOPTFUNCTION using curl_easy_setopt(). We benefited doing this for our workload for S3 cloud in their SDK and feel that if conditions are similar, we need similar implementation for google cloud and its cpp SDK too.

I am not sure how that would help.  The GCS client library only returns control to the application once it received a response from the service, that is, it writes its data *and* waits for the service to respond.  That implies (modulo timeouts) that the socket was flushed.  In other words, there are no "packets" (or bytes) left in any sockets before they are closed and I would expect no effect from turning on `SO_LINGER`.

> 2. Set custom TCP keep intervals and idle times.

That seems useful and we can generalize it, modulo

> 3. Set connection timeouts (CURLOPT_CONNECTTIMEOUT). We have provided our user an option to specify timeouts for other clouds and so we want to consistently honor that for Google cloud too.

Thanks!  We already have a feature request for this (and similar timeouts) #5180, this just adds another reason to implement it.

> 
> So to implement all of these, the options I knew was to get the CURL handle that cloud-cpp SDK is using and add options to it, or create a curl handle and after setting the options, pass it over to SDK to carry it further.

We don't want to expose implementation details (and we look at `libcurl` as an implementation detail) to the surface of the library. That argues against exposing the CURL handles and/or providing any mechanism for you to provide the handles to the library.  This is not just theoretical, there is already a plugin based on gRPC instead of `libcurl`.  While admittedly this is a partial implementation right now, there are good reasons why we would want the library to remain neutral as to the underlying protocol.

> Also, just so you know, we got locked on v0.20.x of google-cloud-cpp last year.

Ack.

**Comment by spnshah** [comment]
> I am not sure how that would help. The GCS client library only returns control to the application once it received a response from the service, that is, it writes its data _and_ waits for the service to respond. That implies (modulo timeouts) that the socket was flushed. In other words, there are no "packets" (or bytes) left in any sockets before they are closed and I would expect no effect from turning on `SO_LINGER`.

This is the situation we run in to currently:
```
# netstat | grep "CLOSE_WAIT" | wc -l
getnameinfo failed
getnameinfo failed
92396

-- From logs --
...google-cloud-cpp/google/cloud/storage/internal/curl_handle.cc:150 ~CurlHandle == curl(Info): Hostname was found in DNS cache
== curl(Info): Hostname in DNS cache was stale, zapped
== curl(Info):   Trying 142.251.33.112...
== curl(Info): Connected to storage.googleapis.com (142.251.33.112) port 443 (#9222)
== curl(Info): successfully set certificate verify locations:
== curl(Info):   CAfile: none
  CApath: /etc/ssl/certs/
== curl(Info): SSLv3, TLS handshake, Client hello (1):
== curl(Info): SSLv3, TLS handshake, Server hello (2):
== curl(Info): SSLv3, TLS handshake, Certificate (11):
== curl(Info): SSLv3, TLS alert, Server hello (2):
== curl(Info): error:02001018:system library:fopen:Too many open files
== curl(Info): Closing connection 9222
```
Am I missing something here from the Client side to close it properly and hence see huge connections in CLOSE_WAIT, or is this something that SDK controls underneath?

> Thanks! We already have a feature request for this (and similar timeouts) #5180, this just adds another reason to implement it.
Any timeframe on when is it planned to get implemented?

**Comment by spnshah** [comment]
Also was able to get high on TIME_WAIT connections, `SO_LINGER` should be able to address this one, I think:
```
Every 1.0s: netstat | grep "TIME_WAIT" | wc -l | tee -a time_conn

getnameinfo failed
5446
```
Is there a way we can limit number of connections?

**Comment by coryan** [comment]
> Is there a way we can limit number of connections?

Yes, the library keeps a pool of connections controlled by:

https://github.com/googleapis/google-cloud-cpp/blob/9ebc00b04b4fd2767c5bee6acdd1172407e6b9eb/google/cloud/storage/client_options.h#L140

The default value is `4 * std::thread::hardware_concurrency()`, if you have a lot of CPUs that can be a high number, but 5400 seems too high (assuming only once process in your machine).

It is possible that #6160 is related, if you have multiple downloads in progress the connection is kept open (and not returned to the pool!) until the `gcs::ObjectReadStream` object is destructed.

>> Thanks! We already have a feature request for this (and similar timeouts) #5180, this just adds another reason to implement it.
>
> Any timeframe on when is it planned?

I will put it in the hopper for Q2.

**Comment by spnshah** [comment]
> Yes, the library keeps a pool of connections controlled by:
> 
> https://github.com/googleapis/google-cloud-cpp/blob/9ebc00b04b4fd2767c5bee6acdd1172407e6b9eb/google/cloud/storage/client_options.h#L140
> 
> The default value is `4 * std::thread::hardware_concurrency()`, if you have a lot of CPUs that can be a high number, but 5400 seems too high (assuming only once process in your machine).

Yes it does seem high. Also I didn't recall when posting, but I was manually computing that and setting connection pool size of `MIN(4 * std::thread::hardware_concurrency(), 64`). I didn't realize it was per process. I will lower it down and check again. I wanted it higher to achieve better performance.

> It is possible that #6160 is related, if you have multiple downloads in progress the connection is kept open (and not returned to the pool!) until the `gcs::ObjectReadStream` object is destructed.

Yes, it does seem like it since we have multiple downloads going on. 

Also, did you got a chance to look at the `CLOSE_WAIT` situation I mentioned above? Will it also be related to more connections or not releasing CURL handle issue?

The following is on a separate note, so if you feel this thread isn't quite suited for below question, I can start a separate one.
 
I have a concern for the parameters which I might need to tweak to get better performance out of google-cloud-cpp. Is there a documentation page explaining what all is available/recommended to get higher throughput from the SDK? 

The reason to ask this is we ran comparison test using `google-cloud-cpp` and `aws-sdk-cpp`, against a storage in Google cloud and from a VM instance in Google cloud. I feel that we might not be fully aware of all the bells and whistles that the google SDK is providing and we can utilize more of the systems resources (IO/CPU/memory). The way we are using the SDK is pretty standard (using `gcs::Client`) and per the online documentation. Any info would be helpful here.

**Comment by coryan** [comment]
> > It is possible that #6160 is related, if you have multiple downloads in progress the connection is kept open (and not returned to the pool!) until the `gcs::ObjectReadStream` object is destructed.
> 
> Yes, it does seem like it since we have multiple downloads going on.

Could you share some information about what is it that your application / service is doing?  If you rather not do that in a public forum, feel free to contact me via email (I work at google.com, same username as GitHub).

> Also, did you got a chance to look at the `CLOSE_WAIT` situation I mentioned above? Will it also be related to more connections or not releasing CURL handle issue?

I would say definitely maybe 🙂 .  There are a few things that are all related to each other, let me see if I can summarize:

- Having a lot of sockets in `CLOSE_WAIT` state means, as you know, that the application has not called `close(2)` or `shutdown(2)` on those sockets even though the peer, in this case the GCS service, has closed their end. Thinking about it there are two ways in which this can happen:
  - The client library keeps a pool of connections, if this pool is large and not being used, you would have many sockets that are in that state. However, you would not be experiencing the "too many file descriptors" in that case because the next download, upload, or RPC would use one of the idle sockets from the pool (closing it and reconnecting if necessary).
  - The client library keeps the socket (really the CURL handle) out of the pool while a download is in progress. If the application is not actively pulling data from one of these file descriptors the service would (eventually) close it and it would get into the `CLOSE_WAIT` state.  

For the second scenario to be a problem you need to have more than `ulimit -n` downloads in progress at a time (typically 1024, but could be different) **and** you need to ignore some of the downloads for minutes.  I guess you could exceed the `/proc/sys/fs/file-max` limit across all the processes, but that is typically very high (like 1/2 million or a million).

This is all to say: I suspect that the high number of `CLOSE_WAIT` sockets are probably correlated to the #6160 problem and/or with the pool size being too large, but we need to discuss your application in a lot more detail to understand how and what to fix in the library.

> 
> The following is on a separate note, so if you feel this thread isn't quite suited for below question, I can start a separate one.

I will try to answer the main questions below, but if that is not enough let's start a separate issue / question.

> 
> I have a concern for the parameters which I might need to tweak to get better performance out of google-cloud-cpp. Is there a documentation page explaining what all is available/recommended to get higher throughput from the SDK?

I would need to understand more about the specific application to make recommendations to tune performance.  Most likely, the defaults should yield close to optimal performance in most scenarios.

In general, I would say that most performance improvements are achieved by making the right calls into the library more than by tuning specific parameters.  A few examples:

Look at this post:

https://medium.com/google-cloud/fast-gcs-downloads-with-c-e5c749bde05e

We get roughly 1GiB/s on a 4 vCPU machine with the default settings.  It is hard to get more than that on that VM model.  The main improvements are (1) parallelize the download across CPUs, and (2) use `.read()` effectively to reduce data copying.

Likewise, when we ran the benchmarks summarized in https://github.com/googleapis/google-cloud-cpp/issues/4298#issuecomment-635620281, we did not find that any of the usual parameters (buffer sizes for example) had as much of an impact on downloads as disabling the MD5 checksums (which are now disabled by default, so don't need to worry about setting that).

For uploads, as noted in https://github.com/googleapis/google-cloud-cpp/issues/2657#issuecomment-519062231 larger buffers do help, but the improvement from 16 MiB to 32 MiB is much higher than from 32 MiB to 64 MiB (or even 128 MiB). While our defaults are (maybe) too conservative, as described here:

https://github.com/googleapis/google-cloud-cpp/blob/9ebc00b04b4fd2767c5bee6acdd1172407e6b9eb/google/cloud/storage/object_stream.h#L207-L213

you don't need to worry about the tunable parameters **if** you give us data in large `.write()` requests: the library uploads with minimal copying in that case (see #4496 and the associated bug #2296 for details).

> 
> The reason to ask this is we ran comparison test using `google-cloud-cpp` and `aws-sdk-cpp`, against a storage in Google cloud and from a VM instance in Google cloud. I feel that we might not be fully aware of all the bells and whistles that the google SDK is providing and we can utilize more of the systems resources (IO/CPU/memory). The way we are using the SDK is pretty standard (using `gcs::Client`) and per the online documentation. Any info would be helpful here.

Some general advice is, which most likely is not news for your:

- Be aware that GCS has high latency to the first byte, good throughput afterwards, and scales well with many simultaneous requests.  That argues for highly parallel requests when possible.
- Be aware that uploads use very little CPU, are and dominated by network I/O.  That argues (again) for many of them in parallel.
- Be aware that a download after it starts receiving data, would consume around 40% of a vCPU for decrypting the stream [1]. 
 That argues for no more than 2.5 downloads per vCPU.

Take the numbers with a grain of salt, i.e., when I say 40% that could be 30% or 50% (but I would be surprised if it was 10% or 90%).

[1]: I blame decryption from looking a callgrind results in the distant past, but I do not remember writing those down, and do not have them handy, my recollection could be wrong.

**Comment by coryan** [comment]
I think the questions in this issue have been answered, please feel free to reopen if you have other questions.

**Comment by spnshah** [comment]
Yes. Thanks for all the help.
